File: Feature Rich Onebox For Emails/src/services/integrationService.ts
import axios from 'axios';
import { config } from '../config/config';
import { Email } from '../models/email';
import { logger } from '../utils/logger';

// Interface for webhook config
interface Webhook {
  id: string;
  url: string;
  events: string[];
  name: string;
  active: boolean;
}

// Interface for notification
interface SlackNotification {
  title: string;
  text: string;
  color?: string;
  email: Email;
}

export class IntegrationService {
  private webhooks: Webhook[] = [];
  private slackWebhookUrl: string;

  constructor() {
    this.slackWebhookUrl = config.slack.webhookUrl;
    
    // Load any existing webhooks from configuration
    // In a production environment, these would come from a database
    const webhookUrl = config.webhook.url;
    if (webhookUrl) {
      this.webhooks.push({
        id: 'default',
        url: webhookUrl,
        events: ['interested_email', 'new_email'],
        name: 'Default Webhook',
        active: true
      });
    }
    
    logger.info(`Integration service initialized with ${this.webhooks.length} webhooks`);
  }

  // Send notification to Slack
  async sendSlackNotification(notification: SlackNotification): Promise<boolean> {
    if (!this.slackWebhookUrl) {
      logger.warn('Slack webhook URL not configured. Cannot send notification.');
      return false;
    }

    try {
      const { title, text, color = 'good', email } = notification;
      
      const payload = {
        attachments: [
          {
            color: color,
            title: title,
            title_link: `mailto:${email.from}?subject=Re: ${email.subject}`,
            text: text,
            fields: [
              {
                title: 'From',
                value: email.from,
                short: true
              },
              {
                title: 'Account',
                value: email.accountId,
                short: true
              },
              {
                title: 'Category',
                value: email.category || 'Uncategorized',
                short: true
              },
              {
                title: 'Received',
                value: email.receivedDate ? new Date(email.receivedDate).toLocaleString() : 'Unknown',
                short: true
              }
            ],
            footer: 'Email Aggregator',
            ts: Math.floor(Date.now() / 1000)
          }
        ]
      };
      
      const response = await axios.post(this.slackWebhookUrl, payload);
      
      if (response.status === 200) {
        logger.info(`Sent Slack notification for email "${email.subject}"`);
        return true;
      } else {
        logger.error(`Error sending Slack notification: ${response.status} ${response.statusText}`);
        return false;
      }
    } catch (error) {
      logger.error('Error sending Slack notification:', error);
      return false;
    }
  }

  // Register a new webhook
  async registerWebhook(webhook: Omit<Webhook, 'id'>): Promise<Webhook> {
    const id = Date.now().toString();
    
    const newWebhook: Webhook = {
      id,
      ...webhook
    };
    
    this.webhooks.push(newWebhook);
    logger.info(`Registered new webhook "${webhook.name}" for events: ${webhook.events.join(', ')}`);
    
    return newWebhook;
  }

  // Trigger webhooks for an event
  async triggerWebhooks(event: string, data: any): Promise<any[]> {
    const results = [];
    
    // Find all active webhooks for this event
    const matchingWebhooks = this.webhooks.filter(webhook => 
      webhook.active && webhook.events.includes(event)
    );
    
    if (matchingWebhooks.length === 0) {
      return [];
    }
    
    logger.info(`Triggering ${matchingWebhooks.length} webhooks for event "${event}"`);
    
    // Call each webhook
    for (const webhook of matchingWebhooks) {
      try {
        const payload = {
          event,
          timestamp: new Date().toISOString(),
          data
        };
        
        const response = await axios.post(webhook.url, payload, {
          headers: {
            'Content-Type': 'application/json',
            'X-Event-Name': event
          },
          timeout: 5000 // 5 second timeout
        });
        
        results.push({
          webhookId: webhook.id,
          success: true,
          statusCode: response.status
        });
        
        logger.info(`Successfully triggered webhook "${webhook.name}" for event "${event}"`);
      } catch (error) {
        logger.error(`Error triggering webhook "${webhook.name}" for event "${event}":`, error);
        
        results.push({
          webhookId: webhook.id,
          success: false,
          error: error.message
        });
      }
    }
    
    return results;
  }

  // Test a webhook
  async testWebhook(webhookId: string, event: string): Promise<any> {
    // Find webhook
    const webhook = this.webhooks.find(wh => wh.id === webhookId);
    
    if (!webhook) {
      return { success: false, error: 'Webhook not found' };
    }
    
    // Test payload
    const testPayload = {
      event,
      timestamp: new Date().toISOString(),
      data: {
        test: true,
        message: `Test payload for event "${event}"`,
        webhookName: webhook.name
      }
    };
    
    try {
      const response = await axios.post(webhook.url, testPayload, {
        headers: {
          'Content-Type': 'application/json',
          'X-Event-Name': event,
          'X-Test': 'true'
        },
        timeout: 5000
      });
      
      logger.info(`Successfully tested webhook "${webhook.name}" for event "${event}"`);
      
      return {
        success: true,
        data: {
          webhook,
          statusCode: response.status,
          response: response.data
        }
      };
    } catch (error) {
      logger.error(`Error testing webhook "${webhook.name}" for event "${event}":`, error);
      
      return {
        success: false,
        error: error.message,
        data: {
          webhook
        }
      };
    }
  }
}

File: Feature Rich Onebox For Emails/services/integration_service.py
import json
import logging
import os
import requests
from datetime import datetime
from models import Webhook, Email
from app import db

logger = logging.getLogger(__name__)

class IntegrationService:
    """Service for handling external integrations like webhooks and Slack notifications"""
    
    def __init__(self):
        self.slack_webhook_url = os.environ.get('SLACK_WEBHOOK_URL')
    
    def trigger_webhooks(self, event, data):
        """Trigger all matching webhooks for an event"""
        try:
            # Find webhooks that are subscribed to this event
            webhooks = Webhook.query.filter(
                Webhook.events.like(f"%{event}%"),
                Webhook.active == True
            ).all()
            
            if not webhooks:
                logger.debug(f"No webhooks found for event {event}")
                return []
                
            results = []
            
            # Call each webhook
            for webhook in webhooks:
                try:
                    # Prepare webhook data
                    webhook_data = {
                        "event": event,
                        "timestamp": datetime.utcnow().isoformat(),
                        "data": data
                    }
                    
                    # Call webhook URL
                    response = requests.post(
                        webhook.url,
                        json=webhook_data,
                        headers={"Content-Type": "application/json"},
                        timeout=5
                    )
                    
                    success = 200 <= response.status_code < 300
                    
                    # Record result
                    results.append({
                        "webhook_id": webhook.id,
                        "name": webhook.name,
                        "success": success,
                        "status_code": response.status_code,
                        "response": response.text[:100]  # Truncate long responses
                    })
                    
                    # Update last triggered time
                    webhook.last_triggered = datetime.utcnow()
                    db.session.commit()
                    
                except Exception as e:
                    logger.error(f"Error calling webhook {webhook.name}: {str(e)}")
                    results.append({
                        "webhook_id": webhook.id,
                        "name": webhook.name,
                        "success": False,
                        "error": str(e)
                    })
            
            return results
            
        except Exception as e:
            logger.error(f"Error triggering webhooks: {str(e)}")
            return []
    
    def test_webhook(self, webhook):
        """Test a webhook with sample data"""
        try:
            # Create test data
            test_data = {
                "event": "test",
                "timestamp": datetime.utcnow().isoformat(),
                "data": {
                    "message": "This is a test webhook from Onebox Email Aggregator",
                    "webhook_id": webhook.id,
                    "webhook_name": webhook.name
                }
            }
            
            # Call webhook URL
            response = requests.post(
                webhook.url,
                json=test_data,
                headers={"Content-Type": "application/json"},
                timeout=5
            )
            
            success = 200 <= response.status_code < 300
            
            return {
                "success": success,
                "status_code": response.status_code,
                "response": response.text[:100]  # Truncate long responses
            }
            
        except Exception as e:
            logger.error(f"Error testing webhook: {str(e)}")
            return {"success": False, "error": str(e)}
    
    def send_slack_notification(self, message, email=None, color="#36a64f"):
        """Send a notification to Slack"""
        if not self.slack_webhook_url:
            logger.debug("Slack webhook URL not configured")
            return False
            
        try:
            # Create Slack message payload
            payload = {
                "attachments": [
                    {
                        "color": color,
                        "pretext": message,
                        "mrkdwn_in": ["text", "pretext"]
                    }
                ]
            }
            
            # Add email details if provided
            if email:
                payload["attachments"][0].update({
                    "title": email.subject,
                    "text": f"From: {email.sender}\nCategory: {email.category or 'uncategorized'}",
                    "footer": f"Email ID: {email.id}"
                })
            
            # Send to Slack
            response = requests.post(
                self.slack_webhook_url,
                json=payload,
                headers={"Content-Type": "application/json"},
                timeout=5
            )
            
            success = 200 <= response.status_code < 300
            
            if not success:
                logger.error(f"Slack notification failed: {response.status_code} {response.text}")
                
            return success
            
        except Exception as e:
            logger.error(f"Error sending Slack notification: {str(e)}")
            return False
    
    def notify_new_email(self, email_id):
        """Send notification for a new email"""
        email = Email.query.get(email_id)
        if not email:
            return False
            
        # Determine message based on category
        category = email.category or "uncategorized"
        color = "#36a64f"  # default green
        
        if category == "interested":
            message = "🔥 New Interested Lead!"
            color = "#36a64f"  # green
        elif category == "meeting_booked":
            message = "📅 New Meeting Booked!"
            color = "#2eb6f2"  # blue
        elif category == "spam":
            message = "🗑️ Spam Email Received"
            color = "#dddddd"  # gray
        elif category == "out_of_office":
            message = "🏖️ Out of Office Reply"
            color = "#dddddd"  # gray
        else:
            message = "📬 New Email Received"
            color = "#f2c744"  # yellow
            
        # Send notification
        return self.send_slack_notification(message, email, color)